Allow chunked restart sampling

blepping opened the pull request:
@ssitu this isn't ready yet, but i wanted to check if you were interested in this approach. 

(note: i don't plan to leave the custom noise node in, but i'd love it if the backend support for custom noise could remain. also always explaining the plan is just temporarily enabled for debug purposes)

the main change is this adds a chunked restart sampling mode which divides the sigmas into segments that can be handed to the sampling function rather than giving it only a single step at a time. i _believe_ this also makes it work more like the a1111 version which keeps state for the second order stuff (though it only supports heun as far as i know). with the current comfy implementation, samplers like dpmpp_2m get fresh state every step, heunpp2 just runs an euler step if the second sigma is the last one (which also would be the case when passing single steps).

with chunked mode disabled it should work exactly the same as the current version. non-chunked mode could be the default and possibly a toggle added to the new custom sampler node.

by the way, in case you're wondering what the custom noise stuff is about - i have a bunch of custom noise stuff here: https://github.com/blepping/ComfyUI-sonar . with the backend support in place, i could import restart sampling as a module and enable a restart sampler node with custom noise input without having to maintain a separate fork. other people could potentially use that i guess, but there's probably no one else crazy enough to want to use non-gaussian noise with restart sampling. :)

ssitu commented:
I've been meaning to do this to see what happens to SDE solvers and if it helps it or not, since they have a BrownianTreeNoiseSampler that doesn't get carried over to the next step because of the single steps. I think SDE solvers might work if we carry over that state in some way, but I haven't looked that far into it. Anyway, it'll be a good addition so that solvers that use more than 1 step will be able to do that most of the time. 
I'm okay with having the noise support for you. I'm sure that there's plenty of people who like experimenting with different noise.

blepping commented:
how would you like to handle stuff like the default setting, having it as an option in samplers, etc?

> to see what happens to SDE solvers and if it helps it or not

it does help quite a bit.

dpmpp_sde non-chunked:

![image](https://github.com/ssitu/ComfyUI_restart_sampling/assets/157360029/f775befe-5678-4d16-bc9d-671ef3b472d2)

chunked:

![image](https://github.com/ssitu/ComfyUI_restart_sampling/assets/157360029/8d58ff10-fbd0-45df-a204-205fe5d2b856)

(not amazing obviously but it's better than pure noise. :)

this wasn't possible before the custom sampler node, but it actually is possible to use SDE samplers in non-chunked mode as long as you turn s_noise way down.

non-chunked, s_noise 0.325:

![image](https://github.com/ssitu/ComfyUI_restart_sampling/assets/157360029/6f1937ea-ea13-4819-be8a-12a73b994a55)

with chunked, around s_noise 0.825 is decent. (testing with an EPS sd1.5 model.)

ssitu commented:
> (not amazing obviously but it's better than pure noise. :)
This wasn't possible before the custom sampler node, but it actually is possible to use SDE samplers in non-chunked mode as long as you turn s_noise way down.

Now I'm wondering if it's any better than just having no restarts because of the deleted state when a restart happens. It's still pretty cool that it's not all noise.

> how would you like to handle stuff like the default setting, having it as an option in samplers, etc?

It should be fine having the setting as a toggle at the bottom of each of the nodes. I think we can default it to being chunked, since that will make multi-step solvers actually function how they're supposed to. If you're up to it, you could measure the difference with some aesthetics metric, completely optional though. 

I like the way you organized the code! I think it can use some comments to help with readability, and I know I'm being a hypocrite for barely having comments in the code I had before 😅. I also like the summary of the plan that prints out before the sampling, it'll give people a better idea of what's going on.

blepping commented:
> Now I'm wondering if it's any better than just having no restarts because of the deleted state when a restart happens.

i'd guess probably not (at least in most cases) at least when using brownian noise. i have a node in my Sonar stuff that lets you override some sampler parameters:

![image](https://github.com/ssitu/ComfyUI_restart_sampling/assets/157360029/e06a5423-5221-4333-9164-79fa8014e745)

so it's possible to do something like that to make `dpmpp_sde` use perlin noise instead which probably cooperates with restart sampling better than brownian noise.

> It should be fine having the setting as a toggle at the bottom of each of the nodes. I think we can default it to being chunked

can do. i added it except for the simple variant (which just uses the default of enabled) assuming you wanted to keep that one simple/less parameters.

> If you're up to it, you could measure the difference with some aesthetics metric, completely optional though.

i don't really know anything about that stuff so unfortunately i'm going to skip that bit.

> I also like the summary of the plan that prints out before the sampling, it'll give people a better idea of what's going on.

it's kind of a lot of stuff to be spewing out to the console though, especially when chunked mode isn't enabled. i made it so it's controlled by an environment variable (documented in the README) for people who want to see that debug info.

>  I know I'm being a hypocrite for barely having comments in the code I had before

not at all, you wrote your code and you know how it works. that's not true for my stuff though, and ultimately you're the one responsible for maintaining it. i added some comments, please let me know if that's what you were looking for or if parts are still unclear. it's hard to predict the information someone else might need.

***

i'm bad about scope creep so i also added some extra stuff in the latest commit:

you can set the segments to two special values: `default` to just use the default segments, and `a1111` to calculate the segments like A1111 does (or at least how i think it works from reading the code). described here: https://github.com/ssitu/ComfyUI_restart_sampling/issues/10#issuecomment-2004423003

this is also documented in the README.

i think this is ready for review now (only lightly tested and as before the disclaimer that i only kind of know what i'm doing still applies. :)

ssitu commented:
Thank you for the comments, and also for solving #10! One thing I'm not so sure about is having "default" as a special input. I intended the default segments to serve as a template/placeholder for people to work off of. What do you think? Was it too long?
As for testing, I'll do some checks to see if everything's working as expected when I get the chance.

blepping commented:
@ssitu 

no problem.

> What do you think? Was it too long?

this one i actually have an opinion on. :) i think keep it is pretty useful. it's not that it's too long, but currently to get back to a known state you need to either have the default segments memorized, recreate the node, or create another sampler node to cut-and-paste the default segments from and then delete it afterward. all of those options are a lot annoying than just entering `default`.

so it's not that it's too long, it's just that it makes resetting the node more convenient after trying something different (or maybe even accidentally changing it). i would say also that realistically 99% of people are just going to use whatever the default is. the majority of people probably also wouldn't even know how to make a reasonable alternative.

***

last pull mainly just has a few cosmetic cleanups. i'm not sure if having chunked mode as a keyword arg to the main node functions is necessary but i _think_ it helps with not messing up existing nodes in workflows that were made before adding the parameter (and it makes more sense to have it set to what the default is, even though it really shouldn't have an effect).

the only functional change there was allowing sigmas to get passed to the main restart sampling function. my thinking was it's a bit of simple infrastructure to possibly at some point make the custom sampler work more like the built in one where you can pass it sigmas.

BTW, if you want i can run the code through formatting either here (though it will make the pull ginormous) or in a separate one after this is resolved. would be cosmetic changes only, but could make the code more readable. some of the lines are very long.

ssitu commented:
> i would say also that realistically 99% of people are just going to use whatever the default is

Yeah I can see that. We can keep how you have it.

> BTW, if you want i can run the code through formatting either here (though it will make the pull ginormous) or in a separate one after this is resolved. would be cosmetic changes only, but could make the code more readable. some of the lines are very long.

I like the idea of having the formatting done in another PR after this one is merged. I'll merge it as soon as I get done with testing it out!

blepping commented:
sounds good! please let me know if you run into any problems/need changes/more explanation/etc.

blepping commented on restart_sampling.py:
```diff
@@ -33,7 +39,25 @@ def resolve_t_value(val, ms):
+            restart_info = "[10,1,0.1,0.2]"
+        else:
+            # Otherwise two restarts with steps // 4 steps.
+            restart_info = f"[{(steps // 4) + 1}, 2, 0.1, 0.2]"
```
ahh, you're right. i did mess that up! thanks for catching it. 

it looks like they're also fixing up the key part (t_min here) with `int(torch.argmin(abs(sigmas - key), dim=0))`, so would i also need to use `int(torch.argmin(abs(sigmas - 0.1), dim=0))` instead there? https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/bef51aed032c0aaa5cfd80445bc4cf0d85b408b5/modules/sd_samplers_extra.py#L50

i don't fully understand what that part is doing.

ssitu replied:
So my understanding is that `max_idx = int(torch.argmin(abs(sigmas - restart_max), dim=0))` is finding the index of the closest sigma to restart_max, which would be 2. 
Then it is used in `sigma_restart = get_sigmas_karras(restart_steps, sigmas[min_idx].item(), sigmas[max_idx].item(), device=sigmas.device)[:-1]`, so ` sigmas[max_idx].item()` is the sigma in the original schedule closest to 2.
So how we have it now, only the `sigma_min`s are rounded to the nearest `sigma` in the schedule as per the paper, and it seems to me that the version in A1111 is rounding their `sigma_max`s to the nearest `sigma`s too.
I'm guessing we can do the same by replacing 2 with the nearest sigma to 2.
Correct me if I'm wrong.

blepping replied:
seems like my initial naive understanding from reading that code was pretty far off. :( sorry about that.

wouldn't we want to choose the closest sigmas for both `0.1` and `0.2` since it's using that calculation in both cases?

ssitu replied:
We already select the closest sigma to 0.1 since it `s_min` rounded to the nearest sigmas in `round_restart_segments`, but we don't do it with `s_max`, which is what it seems like A1111's algorithm does.

blepping replied:
makes sense. how about this? hopefully i got it right this time!

ssitu commented:
Looks good to me! Everything else seems to be working. Is this good to merge?
